rfcs/20210115-tfx-periodic-training.md

```diff
@@ -0,0 +1,445 @@
+generate the new Version of Span 5, Resolver of that pipeline will retrieve
+ExampleGen’s output Spans of day 4 to 6 (not day 3 to 5, Resolver retrieves the
+latest window unless specified) for Trainer. Below shows the graph for backfill
+pipeline:*
+
+<div style="text-align:center"><img src='20210115-tfx-periodic-training/new_version.png', width='500'></div>
+
+Note that the normal run and backfill run need to have the same pipeline name
+for correct ML metadata tracking.
+
+##### Partial pipeline execution
+
+In certain backfill cases users may not want to run the entire pipeline. For
+example, when there are multiple spans in the rolling window that need to be
+backfilled, it’s not necessary to retrain the model multiple times. In such
+cases, user might want to (conceptually) divide the pipeline in two for a faster
+backfill:
+
+*   The first portion contains components like ExampleGen, StatisticsGen and
+    ExampleValidator. This part of the pipeline is scheduled for each Span.
+*   The second portion contains a Resolver plus components like Transform,
+    Trainer, Evaluator and Pusher. This part of the pipeline only needs to run
+    once for updating the model after all Spans are updated.
+
+We plan to provide the ability to specify a partial pipeline to execute at
+runtime (see Future Work below) to enable this workflow.
+
+An alternative approach is to define the two pipeline portions as TFX pipelines
+in their own right, to make backfill easier. This is not recommended because it
+makes the normal case of periodic training more brittle: the two pipelines must
+now be triggered in sequence by some non-TFX mechanism, thereby multiplying the
+possible fault conditions.
+
+##### Training Considerations for Backfill Runs
+
+When there are issues with input data which require new versions of a span, the
+models that have been trained with the problematic data may be of low quality.
+Warmstarting can propagate this damage forward in an unpredictable way. To
+recover a pipeline that uses warmstarting, it’s advisable to retrain from
+scratch. If full retraining is too costly, consider manually picking up a valid
+previous model to warmstart using
+[Importer](https://github.com/tensorflow/tfx/blob/r0.26.0/tfx/components/common_nodes/importer_node.py).
+
+#### ExampleGen Data Selection
+
+As mentioned above, each ExampleGen execution generates only one Span at a time.
+ExampleGen can be configured to process a specific Span (always taking its
+latest Version); if a Span is not specified, ExampleGen selects the input data
+corresponding to the latest Version of the latest Span by default. The latest
+here means the one that has the highest Span id.
+
+Relying on “latest Span” semantics in periodic automation introduces a new
+failure mode: if the data isn’t available as expected the pipeline will
+reprocess old data. While it’s possible to guard against this with upstream
```

Comment:
reprocessing can be avoided with caching. Wondering what's the tradeoff between letting TFX just skip re-processing with caching vs using runtime parameters.

Reply:
"reprocessing can be avoid with caching." yes for the component without input change.
If a component's input depends on mlmd (e.g., resolver), the outcome might change for the reprocessing.
E.g., the latest blessed model will change when you run the pipeline the second time.

besides that, though it's cached without reprocessing, users might not notice there is a data missing happened upstream. Unless there are good alert/monitoring system available at user side, we suggest to specify the span to expose this issue when upstream is missing.